CRM/Mailing/ActionTokens.php

```diff
@@ -0,0 +1,121 @@
+ *   'contact_id' => int, contact_id,
+ *   'email' => string, email
+ *   'phone' => string, phone
+ */
+class CRM_Mailing_ActionTokens extends \Civi\Token\AbstractTokenSubscriber {
+
+  /**
+   * Class constructor.
+   */
+  public function __construct() {
+    // TODO: Think about supporting dynamic tokens like "{action.subscribe.\d+}"
+    parent::__construct('action', array(
+      'subscribeUrl' => ts('Subscribe URL (Action)'),
+      'forward' => ts('Forward URL (Action)'),
+      'optOut' => ts('Opt-Out (Action)'),
+      'optOutUrl' => ts('Opt-Out URL (Action)'),
+      'reply' => ts('Reply (Action)'),
+      'unsubscribe' => ts('Unsubscribe (Action)'),
+      'unsubscribeUrl' => ts('Unsubscribe URL (Action)'),
+      'resubscribe' => ts('Resubscribe (Action)'),
+      'resubscribeUrl' => ts('Resubscribe URL (Action)'),
+      'eventQueueId' => ts('Event Queue ID'),
+    ));
+  }
+
+  /**
+   * Evaluate the content of a single token.
+   *
+   * @param \Civi\Token\TokenRow $row
+   *   The record for which we want token values.
+   * @param string $entity
+   * @param string $field
+   *   The name of the token field.
+   * @param mixed $prefetch
+   *   Any data that was returned by the prefetch().
+   *
+   * @return mixed
+   * @throws \CRM_Core_Exception
+   */
+  public function evaluateToken(
+    \Civi\Token\TokenRow $row,
+    $entity,
+    $field,
+    $prefetch = NULL
+  ) {
+    // Most CiviMail action tokens were implemented via getActionTokenReplacement().
+    // However, {action.subscribeUrl} has a second implementation via
+    // replaceSubscribeInviteTokens(). The two appear mostly the same.
+    // We use getActionTokenReplacement() since it's more consistent. However,
+    // this doesn't provide the dynamic/parameterized tokens of
+    // replaceSubscribeInviteTokens().
+
+    if (empty($row->context['mailingJobId']) || empty($row->context['mailingActionTarget']['hash'])) {
+      throw new \CRM_Core_Exception("Error: Cannot use action tokens unless context defines mailingJobId and mailingActionTarget.");
+    }
+
+    if ($field === 'eventQueueId') {
+      $row->format('text/plain')->tokens($entity, $field, $row->context['mailingActionTarget']['id']);
+      return;
+    }
+
+    list($verp, $urls) = CRM_Mailing_BAO_Mailing::getVerpAndUrls(
+      $row->context['mailingJobId'],
+      $row->context['mailingActionTarget']['id'],
+      $row->context['mailingActionTarget']['hash'],
+      // Note: Behavior is already undefined for SMS/'phone' mailings...
+      $row->context['mailingActionTarget']['email']
+    );
+
+    $row->format('text/plain')->tokens($entity, $field,
+      CRM_Utils_Token::getActionTokenReplacement(
+        $field, $verp, $urls, FALSE));
+    $row->format('text/html')->tokens($entity, $field,
+      CRM_Utils_Token::getActionTokenReplacement(
+        $field, $verp, $urls, TRUE));
+  }
+
+  protected function getTrackOpenUrl(\Civi\Token\TokenRow $row) {
```

Comment:
Good catch. It was a leftover from an earlier/misconceived approach to open-tracking. I'll remove it.